types/react/index.d.ts

```diff
@@ -77,7 +78,12 @@ declare namespace React {
     type ComponentType<P = {}> = ComponentClass<P> | StatelessComponent<P>;
 
     type Key = string | number;
-    type Ref<T> = string | { bivarianceHack(instance: T | null): any }["bivarianceHack"];
+
+    interface RefObject<T> {
+        readonly current: T | null;
+    }
```

Comment:
Would it be possible to use a secret `unique symbol` to brand this interface and make it only constructible via `createRef`? Or would any `unique symbol`s we declare be visible to user code, even if it doesn't exist?

Reply:
I'm not sure what you mean -- do you mean we should add another property in the interface to ensure that users don't try to pass any old object into `ref`?

I tried something like this, which does work, let me know if that works for you:
```tsx
// top level declaration
declare const ReactRef: unique symbol;

declare namespace React {
  // should only be constructed by createRef
  interface RefObject<T> {
    readonly current: T | null;
    [ReactRef]: T
}
```

```tsx
// ReactRef is unable to be imported since it's a non-exported const
import { RefObject, Component, createRef } from 'react'

class MyComponent extends Component {
    mySneakyRef: RefObject<HTMLDivElement> = { // $ExpectError
        current: null
    }
    myTrueRef: RefObject<HTMLDivElement> = createRef()

    render() {
        return (
            <>
                <div ref={this.mySneakyRef} />    
                <div ref={this.myTrueRef} />    
            </>    
        )
    }
}
```

Reply:
Interestingly though, the source really does just construct a plain object and return it (despite the fact that it's sealed in `DEV` mode), so a user passing in a plain object into `ref` will also find that it works:

https://github.com/facebook/react/blob/master/packages/react/src/ReactCreateRef.js

Reply:
That's interesting 🤔 

I thought it'd brand the returned object, similar to the context/provider pair, but I guess they get away with not branding it since it's not something you give as the first argument to `createElement` 😄